tests/system/framework/roles/shadow.py

```diff
@@ -0,0 +1,161 @@
+"""shadow multihost role."""
+
+from __future__ import annotations
+
+import shlex
+from typing import Dict
+
+from pytest_mh.conn import ProcessLogLevel, ProcessResult
+
+from ..hosts.shadow import ShadowHost
+from .base import BaseLinuxRole
+
+__all__ = [
+    "Shadow",
+]
+
+
+class Shadow(BaseLinuxRole[ShadowHost]):
+    """
+    shadow role.
+
+    Provides unified Python API for managing and testing shadow.
+    """
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+
+        self._users: set[str] = set()
+        self._groups: set[str] = set()
+
+    def teardown(self) -> None:
+        """
+        Clean up the environment.
+        """
+        cmd = ""
+
+        if self._users:
+            cmd += "\n".join([f"userdel '{x}' --force --remove" for x in self._users])
+            cmd += "\n"
+        if self._groups:
+            cmd += "\n".join([f"groupdel '{x}' --force" for x in self._groups])
+            cmd += "\n"
```

Comment:
It may not be the best idea to call tool that you are actually testing in teardown. I think you should backup/restore the files that shadow-utils is touching (/etc/passwd,groups and /etc/home?), perhaps backup/restore in the host class is sufficient?

Reply:
Do you mean `ShadowHost`? I'm already doing the backup, except for /home, would that be enough?

Reply:
Done. `ShadowHost` takes care of backup and restore for each test case